### PR TITLE
Fix calling isinstance on torch.nn.Parameter

### PR DESCRIPTION
**Pull request by yanboliang**
Fix this bug:
```
def fn(x):
    a = torch.nn.Parameter(torch.rand(2, 3))
    if isinstance(a, torch.Tensor):
        return x + 1
    else:
        return x - 1

x = torch.tensor([2.5])
print(fn(x))

with torchdynamo.optimize("eager"):
    print(fn(x))
```
Pytorch output: ```[3.5]```
TorchDynamo output: ```[1.5]```

### CODE REVIEW

File: torchdynamo/variables/tensor.py [hunk 1]
```diff
@@ -311,7 +311,10 @@ def call_isinstance(self, tensor_type):
 
         def check_type(ty):
             if ty not in tensortype_to_dtype:
-                return self.python_type() is ty
+                if self.python_type() is torch.nn.Parameter and ty is torch.Tensor:
```
Comment (voznesenskym) [inline torchdynamo/variables/tensor.py:314]: We already have utils is_tensor:

```
def istensor(obj):
    """Check of obj is a tensor"""
    tensor_list = (
        torch.Tensor,
        torch.nn.Parameter,
        *config.traceable_tensor_subclasses,
    )
    if fake_tensors_available:
        tensor_list = tensor_list + (torch._subclasses.FakeTensor,)
    return istype(obj, tensor_list)
```

Can this share logic with that?

Comment (yanboliang) [reply torchdynamo/variables/tensor.py:314]: No, I don't think sharing the logic can simplify code. Can you take a look at the complete implementation of ```call_isinstance```? They are different purpose.

Comment (voznesenskym) [reply torchdynamo/variables/tensor.py:314]: Why do we have two different valid sets of subclasses for torch tensor? Is that full list not valid? Which elements of that list do we not want to support?

Comment (yanboliang) [reply torchdynamo/variables/tensor.py:314]: Can you take a look at the complete implementation of ```call_isinstance```? It's using ```dtype``` to decide the tensor type. I think here the only problem is we don't handle ```torch.nn.Parameter``` well. For other tensor type listed at ```istensor```, we already handle them well. Or if you can suggest a better and simplified way to reuse them?

Comment (voznesenskym) [reply torchdynamo/variables/tensor.py:314]:
```
def valid_tensor_subtypes():
    tensor_list = (
        torch.Tensor,
        torch.nn.Parameter,
        *config.traceable_tensor_subclasses,
    )
    if fake_tensors_available:
        tensor_list = tensor_list + (torch._subclasses.FakeTensor,)
    return tensor_list
```

```
def istensor(obj):
    """Check of obj is a tensor"""
    return istype(obj, valid_tensor_subtypes())
```

```
def check_type(ty):
            if ty not in tensortype_to_dtype:
                if self.python_type() in valid_tensor_subtypes():
                  ...
```

Comment (voznesenskym) [reply torchdynamo/variables/tensor.py:314]: Meta comment - please stop resolving the conversations until the concern is addressed!

Comment (yanboliang) [reply torchdynamo/variables/tensor.py:314]: Can you look at [these code](https://github.com/pytorch/torchdynamo/blob/main/torchdynamo/variables/tensor.py#L161), ```self.python_type()``` already return ```torch.Tensor``` if it's a ```FakeTensor```, why we need to add redundant code?
I'd recommend you to take a closer look at ```TensorVariable```, actually for regular tensor, fake tensor and other tensors, their python type is ```torch.Tensor```, only parameter tensor returns ```torch.nn.Parameter```, that' what this PR fixed? Thank you!

Comment (voznesenskym) [reply torchdynamo/variables/tensor.py:314]: Because if we ever get a new tensor subtype, we would want to support it in both places, right?

Comment (yanboliang) [reply torchdynamo/variables/tensor.py:314]: No, if there is a new tensor subtype, we should not update where I changed, we should do some update like [this line](https://github.com/pytorch/torchdynamo/blob/main/torchdynamo/variables/tensor.py#L161) to make its ```class_type``` is ```torch.Tensor```.

Comment (anijain2305) [inline torchdynamo/variables/tensor.py:314]: use `isinstance(self.python_type(), ty)`

Comment (yanboliang) [reply torchdynamo/variables/tensor.py]: sync up with @anijain2305 offline and find ```issubclass``` is the most concise way.

**Comment by anijain2305** [comment]
@voznesenskym Can you PTAL?

**Review by anijain2305 — APPROVED** [review]

**Pull request merged**